### PR TITLE
BUG: fixed regression in np.histogram which caused input floating-point values to be modified

### PR DESCRIPTION
**Pull request by astrofrog**
@jaimefrio @rgommers @nayyarv - this should solve the issue described in #6120 related to in-place modifications of the input array, but I think @jaimefrio is also right that the original test needs fixing too.

This results in a 5-7% loss in performance, but this is very little compared to the original factor of 10x performance gain, so nothing to worry about.

Sorry for introducing this bug!

### CODE REVIEW

File: numpy/lib/function_base.py [hunk 1]
```diff
@@ -232,7 +232,7 @@ def histogram(a, bins=10, range=None, normed=False, weights=None,
                 tmp_a = tmp_a[keep]
                 if tmp_w is not None:
                     tmp_w = tmp_w[keep]
-            tmp_a = np.array(tmp_a, dtype=float, copy=False)
+            tmp_a = np.array(tmp_a, dtype=float, copy=True)
```
Comment (jaimefrio) [inline numpy/lib/function_base.py]: I think the intention of that line is better conveyed using the `astype` method, which defaults to `copy=True`, probably not worth setting it explicitly. Also, I have a feeling it is better to use `np.double` or `np.float64` rather than `float`, but perhaps @njsmith has some thoughts on the proper thing to use from the recent discussion on deprecating `np.float` and friends.

**Comment by jaimefrio** [comment]
I was kind of hoping that, since we have to make a copy anyway, we could get rid of the check on the entries of `keep` being all `True`. Haven't thought about it hard enough, but there doesn't seem to be any obvious way to get that to happen...

**Comment by astrofrog** [comment]
@jaimefrio - I've changed to using ``astype``. I'm also happy to change the actual type to ``np.float64`` is you want, so just let me know.

**Comment by nayyarv** [comment]
Looks good, @astrofrog, the new test is definitely warranted.

For large a.size, would there be an advantage to preallocating the tmp_a space instead of making a copy each time. i.e. something like

```python
tmp_a_storage = np.zeros(min(BLOCK, len(a)))
# ... somewhat inelegant off the cuff hack
if not np.logical_and.reduce(keep):
    tmp_a_storage[keep] = tmp_a[keep]
    tmp_a = tmp_a_storage[keep]
else:
    tmp_a_storage = tmp_a[:]
    tmp_a = tmp_a_storage
```
This changes tmp_a to point to your pre-allocated storage which does not need to be copied each time
This way you're not constantly allocating and freeing memory BLOCKs for large a. From my last experiments with timeit, it's the memory overhead that should be causing the 5-7% loss of efficiency, not the computational time. 

I also agree that scipy should update their test, or get rid of it entirely. Edge case testing with floats is rarely a good idea.

File: numpy/lib/function_base.py (hunk 1 above)
Comment (njsmith) [inline numpy/lib/function_base.py]: No particular opinion on float versus np.float64. They do the same thing,
and I can't imagine this ever changing.

**Comment by jaimefrio** [comment]
I'm merging this as is to fix scipy's issues. If there are improvements to be done elsewhere, a new PR is wecome, @nayyarv! Thanks, @astrofrog.
